src/backend/distributed/commands/cascade_citus_table_function.c

```diff
@@ -0,0 +1,366 @@
+	 * table functions already have their own logics to handle partition relations.
+	 */
+	List *nonPartitionRelationIdList =
+		RemovePartitionRelationIds(fKeyConnectedRelationIdList);
+
+	/*
+	 * Our foreign key subgraph can have distributed tables which might already
+	 * be modified in current transaction. So switch to sequential execution
+	 * before executing any ddl's to prevent erroring out later in this function.
+	 */
+	EnsureSequentialModeForCitusTableCascadeFunction(nonPartitionRelationIdList);
+
+	/* store foreign key creation commands before dropping them */
+	List *fKeyCreationCommands =
+		GetFKeyCreationCommandsForRelationIdList(nonPartitionRelationIdList);
+
+	/*
+	 * Note that here we only drop referencing foreign keys for each relation.
+	 * This is because referenced foreign keys are already captured as other
+	 * relations' referencing foreign keys.
+	 */
+	DropRelationIdListForeignKeys(nonPartitionRelationIdList);
+	ExecuteCitusTableFunctionForRelationIdList(nonPartitionRelationIdList,
+											   executeCitusTableFunctionFlag);
+
+	/* now recreate foreign keys on tables */
+	ExecuteAndLogDDLCommandList(fKeyCreationCommands);
+}
+
+
+/*
+ * LockRelationsWithLockMode sorts given relationIdList and then acquires
+ * specified lockMode on those relations.
+ */
+static void
+LockRelationsWithLockMode(List *relationIdList, LOCKMODE lockMode)
+{
+	Oid relationId;
+	relationIdList = SortList(relationIdList, CompareOids);
+	foreach_oid(relationId, relationIdList)
+	{
+		LockRelationOid(relationId, lockMode);
+	}
+}
+
+
+/*
+ * RemovePartitionRelationIds returns a list of relation id's by removing
+ * partition relation id's from given relationIdList.
+ */
+static List *
+RemovePartitionRelationIds(List *relationIdList)
+{
+	List *nonPartitionRelationIdList = NIL;
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
+	{
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		nonPartitionRelationIdList = lappend_oid(nonPartitionRelationIdList, relationId);
+	}
+
+	return nonPartitionRelationIdList;
+}
+
+
+/*
+ * EnsureSequentialModeForCitusTableCascadeFunction switches to sequential
+ * execution mode if needed. If it's not possible, then errors out.
+ */
+static void
+EnsureSequentialModeForCitusTableCascadeFunction(List *relationIdList)
+{
+	if (!RelationIdListHasReferenceTable(relationIdList))
+	{
+		/*
+		 * We don't need to switch to sequential execution if there is no
+		 * reference table in our foreign key subgraph.
+		 */
+		return;
+	}
+
+	if (ParallelQueryExecutedInTransaction())
+	{
+		ereport(ERROR, (errmsg("cannot execute table udf because there was a parallel "
+							   "operation on a distributed table in transaction"),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
+	}
+
+	ereport(DEBUG1, (errmsg("switching to sequential query execution mode")));
```

Comment:
maybe a little more details:
```
switching to sequential query execution mode because the operation cascades into distributed tables with foreign keys to reference tables
```